### PR TITLE
Jb docker reorg

### PR DESCRIPTION
**Pull request by jaybdub**

### CODE REVIEW

**Comment by tokk-nv** [comment]
Are we going to count on `--restart always` to also bring containers up upon system boot?
If so, we need to test how that behaves when no user is configured to auto login to the system upon boot.

**Comment by jaybdub** [comment]
That's a good point.  Actually, how does the current service work when auto-login is disabled?  I've always done auto-login.

**Comment by jaybdub** [comment]
Just tested.  It seems like the docker services don't start on boot, or login.

But after I type

```bash
sudo docker ps
```

The containers start, and are then listed.  Seems like the daemon is triggered by ``docker`` commands

**Comment by tokk-nv** [comment]
Hmm, that seems to be a different test result than I have.
It was under my system where user `jetson` auto logins, but when I specify `--restart unless-stopped`, the container came up automatically upon system reboot.

**Comment by jaybdub** [comment]
Seems like on my system the docker daemon is disabled at boot

```
 systemctl list-unit-files
```

image

After doing

```
sudo systemctl enable docker
```

And rebooting, the docker containers are launched.

**Comment by jaybdub** [comment]
I disabled auto-login following this guide.

https://techpiezo.com/linux/enable-or-disable-automatic-login-in-ubuntu-20-04/

Rebooted, and containers are launched.  

I think we're good-to-go, but will want to document or add

```bash
sudo systemctl enable docker
``` 

Somewhere to be safe.

**Comment by jaybdub** [comment]
Only remaining nit.  It's pretty annoying to edit files in jupyter on container, and they have different permissions as host user.

So you need to then use ``sudo`` on host to access those files.  

```
--user $(id -u):$(id -g)
```

Going to test adding the ``--user`` flag in enable scripts.  This way permissions will be the same.

File: docker/base/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+FROM ${BASE_IMAGE}
+
+ENV DEBIAN_FRONTEND=noninteractive
+
+# ====================
+# INSTALL COMMON TOOLS
+# ====================
+RUN apt-get update && apt-get install -y net-tools vim cmake
+
+
```
Comment (tokk-nv) [inline docker/base/Dockerfile:11]: Do you want to add pre-caching of torchvision models?

```
#
# pre-cache torchvision models
#
RUN python3 -c "import torchvision; \
                model = torchvision.models.alexnet(pretrained=True); \
                model = torchvision.models.squeezenet1_1(pretrained=True); \
                model = torchvision.models.resnet18(pretrained=True); \
                model = torchvision.models.resnet34(pretrained=True)"
```

File: docker/base/run.sh [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+    --runtime nvidia \
+    --network host \
+    --privileged \
+    --device /dev/video* \
+    --volume /dev/bus/usb:/dev/bus/usb \
+    --volume /tmp/argus_socket:/tmp/argus_socket \
+    -p 8888:8888 \
+    -v $WORKSPACE:/workspace \
+    --workdir /workspace \
+    jetbot-base:$JETBOT_VERSION
```
Comment (tokk-nv) [inline docker/base/run.sh:13]: Do you want to add `--name` option for the base as well? (I see all the other 3 containers are given name)

File: docker/base/Dockerfile (hunk 1 above)
Comment (jaybdub) [reply docker/base/Dockerfile:11]: Yes these seems good.

I've added this as dependency of Jupyter container.

**Comment by tokk-nv** [comment]
BTW, this error has been happening with jetbot. (Not particular to Docker I believe.)

```
Step 33/33 : RUN cd ${JETBOT_REPO_DIR} && python3 setup.py install
 ---> Running in de57e3758b4c
-- The C compiler identification is GNU 7.5.0
-- The CXX compiler identification is GNU 7.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - not found
-- Looking for pthread_create in pthreads
-- Looking for pthread_create in pthreads - not found
-- Looking for pthread_create in pthread
-- Looking for pthread_create in pthread - found
-- Found Threads: TRUE
-- Found CUDA: /usr/local/cuda (found version "10.2")
-- Configuring done
-- Generating done
-- Build files have been written to: /opt/jetbot
Scanning dependencies of target ssd_tensorrt
[ 50%] Building CXX object jetbot/ssd_tensorrt/CMakeFiles/ssd_tensorrt.dir/FlattenConcat.cpp.o
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:23:17: error: 'nvinfer1' is not a namespace-name
 using namespace nvinfer1;
                 ^~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:23:25: error: expected namespace-name before ';' token
 using namespace nvinfer1;
                         ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:34:1: error: expected class-name before '{' token
 {
 ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:69:5: error: 'Dims' does not name a type; did you mean 'time'?
     Dims getOutputDimensions(int index, const Dims* inputs, int nbInputDims) override
     ^~~~
     time
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:113:82: error: 'cudaStream_t' has not been declared
     int enqueue(int batchSize, const void* const* inputs, void** outputs, void*, cudaStream_t stream) override
                                                                                  ^~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:168:36: error: 'Dims' does not name a type; did you mean 'time'?
     void configureWithFormat(const Dims* inputs, int nbInputs, const Dims* outputDims, int nbOutputs, nvinfer1::DataType type, nvinfer1::PluginFormat format, int maxBatchSize) override
                                    ^~~~
                                    time
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:168:70: error: 'Dims' does not name a type; did you mean 'time'?
     void configureWithFormat(const Dims* inputs, int nbInputs, const Dims* outputDims, int nbOutputs, nvinfer1::DataType type, nvinfer1::PluginFormat format, int maxBatchSize) override
                                                                      ^~~~
                                                                      time
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:168:103: error: 'nvinfer1' has not been declared
     void configureWithFormat(const Dims* inputs, int nbInputs, const Dims* outputDims, int nbOutputs, nvinfer1::DataType type, nvinfer1::PluginFormat format, int maxBatchSize) override
                                                                                                       ^~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:168:122: error: expected ',' or '...' before 'type'
     void configureWithFormat(const Dims* inputs, int nbInputs, const Dims* outputDims, int nbOutputs, nvinfer1::DataType type, nvinfer1::PluginFormat format, int maxBatchSize) override
                                                                                                                          ^~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:198:25: error: 'DataType' has not been declared
     bool supportsFormat(DataType type, PluginFormat format) const override
                         ^~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:198:40: error: 'PluginFormat' has not been declared
     bool supportsFormat(DataType type, PluginFormat format) const override
                                        ^~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:209:5: error: 'IPluginV2' does not name a type
     IPluginV2* clone() const override
     ^~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:245:5: error: 'cublasHandle_t' does not name a type
     cublasHandle_t mCublas;
     ^~~~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:63:9: error: 'int FlattenConcat::getNbOutputs() const' marked 'override', but does not override
     int getNbOutputs() const override
         ^~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:92:9: error: 'int FlattenConcat::initialize()' marked 'override', but does not override
     int initialize() override
         ^~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:100:10: error: 'void FlattenConcat::terminate()' marked 'override', but does not override
     void terminate() override
          ^~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:107:12: error: 'size_t FlattenConcat::getWorkspaceSize(int) const' marked 'override', but does not override
     size_t getWorkspaceSize(int maxBatchSize) const override
            ^~~~~~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:113:9: error: 'int FlattenConcat::enqueue(int, const void* const*, void**, void*, int)' marked 'override', but does not override
     int enqueue(int batchSize, const void* const* inputs, void** outputs, void*, cudaStream_t stream) override
         ^~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:135:12: error: 'size_t FlattenConcat::getSerializationSize() const' marked 'override', but does not override
     size_t getSerializationSize() const override
            ^~~~~~~~~~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:144:10: error: 'void FlattenConcat::serialize(void*) const' marked 'override', but does not override
     void serialize(void* buffer) const override
          ^~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:168:10: error: 'void FlattenConcat::configureWithFormat(const int*, int, const int*, int, int)' marked 'override', but does not override
     void configureWithFormat(const Dims* inputs, int nbInputs, const Dims* outputDims, int nbOutputs, nvinfer1::DataType type, nvinfer1::PluginFormat format, int maxBatchSize) override
          ^~~~~~~~~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:198:10: error: 'bool FlattenConcat::supportsFormat(int, int) const' marked 'override', but does not override
     bool supportsFormat(DataType type, PluginFormat format) const override
          ^~~~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:203:17: error: 'const char* FlattenConcat::getPluginType() const' marked 'override', but does not override
     const char* getPluginType() const override { return FLATTENCONCAT_PLUGIN_NAME; }
                 ^~~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:205:17: error: 'const char* FlattenConcat::getPluginVersion() const' marked 'override', but does not override
     const char* getPluginVersion() const override { return FLATTENCONCAT_PLUGIN_VERSION; }
                 ^~~~~~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:207:10: error: 'void FlattenConcat::destroy()' marked 'override', but does not override
     void destroy() override {}
          ^~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:214:10: error: 'void FlattenConcat::setPluginNamespace(const char*)' marked 'override', but does not override
     void setPluginNamespace(const char* pluginNamespace) override
          ^~~~~~~~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:219:17: error: 'const char* FlattenConcat::getPluginNamespace() const' marked 'override', but does not override
     const char* getPluginNamespace() const override
                 ^~~~~~~~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp: In member function 'int FlattenConcat::initialize()':
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:96:29: error: 'mCublas' was not declared in this scope
         CHECK(cublasCreate(&mCublas));
                             ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:15:21: note: in definition of macro 'CHECK'
         auto ret = (status);                      \
                     ^~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:96:15: error: 'cublasCreate' was not declared in this scope
         CHECK(cublasCreate(&mCublas));
               ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:15:21: note: in definition of macro 'CHECK'
         auto ret = (status);                      \
                     ^~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp: In member function 'void FlattenConcat::terminate()':
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:104:29: error: 'mCublas' was not declared in this scope
         CHECK(cublasDestroy(mCublas));
                             ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:15:21: note: in definition of macro 'CHECK'
         auto ret = (status);                      \
                     ^~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:104:15: error: 'cublasDestroy' was not declared in this scope
         CHECK(cublasDestroy(mCublas));
               ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:15:21: note: in definition of macro 'CHECK'
         auto ret = (status);                      \
                     ^~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp: In member function 'int FlattenConcat::enqueue(int, const void* const*, void**, void*, int)':
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:125:35: error: 'mCublas' was not declared in this scope
                 CHECK(cublasScopy(mCublas, mFlattenedInputSize[i],
                                   ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:15:21: note: in definition of macro 'CHECK'
         auto ret = (status);                      \
                     ^~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:125:23: error: 'cublasScopy' was not declared in this scope
                 CHECK(cublasScopy(mCublas, mFlattenedInputSize[i],
                       ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:15:21: note: in definition of macro 'CHECK'
         auto ret = (status);                      \
                     ^~~~~~
In file included from /usr/include/c++/7/cassert:44:0,
                 from /opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:2:
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp: In member function 'void FlattenConcat::configureWithFormat(const int*, int, const int*, int, int)':
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:183:30: error: request for member 'nbDims' in '*(inputs + ((sizetype)(((long unsigned int)i) * 4)))', which is of non-class type 'const int'
             assert(inputs[i].nbDims == 3);
                              ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:188:34: error: request for member 'd' in '*(inputs + ((sizetype)(((long unsigned int)i) * 4)))', which is of non-class type 'const int'
                 assert(inputs[i].d[dim] == inputs[0].d[dim]);
                                  ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:188:54: error: request for member 'd' in '* inputs', which is of non-class type 'const int'
                 assert(inputs[i].d[dim] == inputs[0].d[dim]);
                                                      ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:192:37: error: request for member 'd' in '*(inputs + ((sizetype)(((long unsigned int)i) * 4)))', which is of non-class type 'const int'
             inputVolume = inputs[i].d[0] * inputs[i].d[1] * inputs[i].d[2];
                                     ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:192:54: error: request for member 'd' in '*(inputs + ((sizetype)(((long unsigned int)i) * 4)))', which is of non-class type 'const int'
             inputVolume = inputs[i].d[0] * inputs[i].d[1] * inputs[i].d[2];
                                                      ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:192:71: error: request for member 'd' in '*(inputs + ((sizetype)(((long unsigned int)i) * 4)))', which is of non-class type 'const int'
             inputVolume = inputs[i].d[0] * inputs[i].d[1] * inputs[i].d[2];
                                                                       ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp: In member function 'bool FlattenConcat::supportsFormat(int, int) const':
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:200:25: error: 'DataType' has not been declared
         return (type == DataType::kFLOAT && format == PluginFormat::kNCHW);
                         ^~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:200:55: error: 'PluginFormat' has not been declared
         return (type == DataType::kFLOAT && format == PluginFormat::kNCHW);
                                                       ^~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp: At global scope:
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:253:1: error: expected class-name before '{' token
 {
 ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:267:11: error: 'PluginFieldCollection' does not name a type
     const PluginFieldCollection* getFieldNames() override { return &mFC; }
           ^~~~~~~~~~~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:269:5: error: 'IPluginV2' does not name a type
     IPluginV2* createPlugin(const char* name, const PluginFieldCollection* fc) override
     ^~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:274:5: error: 'IPluginV2' does not name a type
     IPluginV2* deserializePlugin(const char* name, const void* serialData, size_t serialLength) override
     ^~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:291:12: error: 'PluginFieldCollection' does not name a type
     static PluginFieldCollection mFC;
            ^~~~~~~~~~~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:292:24: error: 'PluginField' was not declared in this scope
     static std::vector<PluginField> mPluginAttributes;
                        ^~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:292:35: error: template argument 1 is invalid
     static std::vector<PluginField> mPluginAttributes;
                                   ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:292:35: error: template argument 2 is invalid
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:263:17: error: 'const char* FlattenConcatPluginCreator::getPluginName() const' marked 'override', but does not override
     const char* getPluginName() const override { return FLATTENCONCAT_PLUGIN_NAME; }
                 ^~~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:265:17: error: 'const char* FlattenConcatPluginCreator::getPluginVersion() const' marked 'override', but does not override
     const char* getPluginVersion() const override { return FLATTENCONCAT_PLUGIN_VERSION; }
                 ^~~~~~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:280:10: error: 'void FlattenConcatPluginCreator::setPluginNamespace(const char*)' marked 'override', but does not override
     void setPluginNamespace(const char* pluginNamespace) override
          ^~~~~~~~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:285:17: error: 'const char* FlattenConcatPluginCreator::getPluginNamespace() const' marked 'override', but does not override
     const char* getPluginNamespace() const override
                 ^~~~~~~~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp: In constructor 'FlattenConcatPluginCreator::FlattenConcatPluginCreator()':
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:257:9: error: 'mFC' was not declared in this scope
         mFC.nbFields = 0;
         ^~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp: At global scope:
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:296:1: error: 'PluginFieldCollection' does not name a type
 PluginFieldCollection FlattenConcatPluginCreator::mFC{};
 ^~~~~~~~~~~~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:297:13: error: 'PluginField' was not declared in this scope
 std::vector<PluginField> FlattenConcatPluginCreator::mPluginAttributes;
             ^~~~~~~~~~~
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:297:24: error: template argument 1 is invalid
 std::vector<PluginField> FlattenConcatPluginCreator::mPluginAttributes;
                        ^
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:297:24: error: template argument 2 is invalid
/opt/jetbot/jetbot/ssd_tensorrt/FlattenConcat.cpp:299:53: error: expected constructor, destructor, or type conversion before ';' token
 REGISTER_TENSORRT_PLUGIN(FlattenConcatPluginCreator);
                                                     ^
jetbot/ssd_tensorrt/CMakeFiles/ssd_tensorrt.dir/build.make:62: recipe for target 'jetbot/ssd_tensorrt/CMakeFiles/ssd_tensorrt.dir/FlattenConcat.cpp.o' failed
make[2]: *** [jetbot/ssd_tensorrt/CMakeFiles/ssd_tensorrt.dir/FlattenConcat.cpp.o] Error 1
CMakeFiles/Makefile2:103: recipe for target 'jetbot/ssd_tensorrt/CMakeFiles/ssd_tensorrt.dir/all' failed
make[1]: *** [jetbot/ssd_tensorrt/CMakeFiles/ssd_tensorrt.dir/all] Error 2
make: *** [all] Error 2
Makefile:83: recipe for target 'all' failed
```

**Pull request merged**